File: app/schemas/admin_status.py
"""Schemas for the admin system status endpoint (Task G.1).

Expone el último resultado del ProviderCanaryJob y un ping de Redis
para que el dashboard/ops pueda verificar salud sin leer logs.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class JobMetricsRead(BaseModel):
    """Snapshot of a registered job's runtime metrics (Task G.4)."""

    name: str
    interval: int
    status: str
    last_execution: datetime | None = None
    next_execution: datetime | None = None
    last_duration: float = 0.0
    execution_count: int = 0
    success_count: int = 0
    failure_count: int = 0
    consecutive_failures: int = 0


class ProviderCanaryStatus(BaseModel):
    """Último snapshot del ProviderCanaryJob.

    Todos los campos son ``None`` si aún no se ha ejecutado el canary
    (p. ej. justo después de reiniciar el proceso).
    """

    success: bool | None = None
    message: str | None = None
    finished_at: str | None = None
    autoscout24: dict | None = None
    mobile_de: dict | None = None
    strict_mobile: bool | None = None
    mobile_status: str | None = None


class ProviderInfo(BaseModel):
    """Resumen de un provider registrado."""

    name: str
    source: str | None = None


class ProvidersStatus(BaseModel):
    """Snapshot del registry de providers y flags ES relevantes."""

    providers: list[str] = []
    default_import_cost_profile: str = ""
    enable_es_market_fixture: bool = False
    enable_coches_net_fixture: bool = False
    enable_autoscout24_es: bool = False


class AdminSystemStatus(BaseModel):
    """Estado del sistema para admin: Redis + último canary + jobs."""

    redis_ok: bool | None = None
    canary: ProviderCanaryStatus
    jobs: list[JobMetricsRead] = []  # nuevo (G.4)
    providers: ProvidersStatus = ProvidersStatus()  # ADMIN.1

File: frontend/src/app/services/adminStatus.ts
import { api } from "@/app/services/api/client";

export type ProviderCanaryStatus = {
  success: boolean | null;
  message: string | null;
  finished_at: string | null;
  autoscout24: Record<string, unknown> | null;
  mobile_de: Record<string, unknown> | null;
  strict_mobile: boolean | null;
  mobile_status: string | null;
};

export type JobMetricsRead = {
  name: string;
  interval: number;
  status: string;
  last_execution: string | null;
  next_execution: string | null;
  last_duration: number;
  execution_count: number;
  success_count: number;
  failure_count: number;
  consecutive_failures: number;
};

export type ProvidersStatus = {
  providers: string[];
  default_import_cost_profile: string;
  enable_es_market_fixture: boolean;
  enable_coches_net_fixture: boolean;
  enable_autoscout24_es: boolean;
};

export type AdminSystemStatus = {
  redis_ok: boolean | null;
  canary: ProviderCanaryStatus;
  jobs: JobMetricsRead[];
  providers: ProvidersStatus;
};

export async function fetchAdminStatus(): Promise<AdminSystemStatus> {
  const { data } = await api.get<AdminSystemStatus>("/admin/status");
  return data;
}

export async function runProviderCanary(): Promise<AdminSystemStatus> {
  // Canary puede tardar (red real); timeout largo solo en esta llamada
  const { data } = await api.post<AdminSystemStatus>(
    "/admin/status/canary",
    null,
    { timeout: 120_000 }
  );
  return data;
}